src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumWatcher.kt

```diff
@@ -429,16 +494,85 @@ class ElectrumWatcher(val client: ElectrumClient, val scope: CoroutineScope) : C
                     )
                     is NotifyWatch -> {
                         if (action.broadcastNotification)
-                            watchNotificationsChannel.send(action.watchEvent)
+                            _notificationsFlow.emit(NotifyWatchEvent(action.watchEvent))
                         else
                             eventChannel.send(ReceiveWatchEvent(action.watchEvent))
                     }
-                    is NotifyTxWithMeta -> txNotificationsChannel.send(action.channelId to action.txWithMeta)
+                    is NotifyTxWithMeta -> {
+                       _notificationsFlow.emit(NotifyTxEvent(action.channelId, action.txWithMeta))
+                    }
                 }
             }
         }
     }
 
+    private fun startTimer() {
+        if (timerJob != null) return
+
+        val timeMillis: Long = 2 * 1_000 // fire timer every 2 seconds
+        timerJob = launch {
+            delay(timeMillis)
+            while(isActive) {
+                checkIsUpToDate()
+                delay(timeMillis)
+            }
+        }
+    }
+
+    private fun stopTimer() {
+        timerJob?.cancel()
+        timerJob = null
+    }
+
+    private suspend fun checkIsUpToDate(): Unit {
+
+        // Get a list of timestamps from the client for all outgoing requests.
+        // This returns an array of `RequestResponseTimestamp` instances, which includes:
+        // - id: Int
+        // - request: ElectrumRequest
+        // - lastResponseTimestamp: Long?
+        //
+        // If lastResponseTimestamp is null, we have an outstanding request pending.
+        //
+        val info = client.requestResponseTimestamps()
+        if (info == null) {
+            // Client isn't connected (i.e. client.state !is ClientRunning)
+            return
+        }
+
+        // We want to emit an "up-to-date" notification to the client.
+        // This is used during WatchTower operation, where the client connects to electrum
+        // in order to ensure the acinq server didn't publish an old commit (i.e. attempt to cheat).
+        //
+        // So we want to emit this message when we're fairly confident the electrum server
+        // has delivered all the pending information we need. Here's how we do it:
+        // - we track the last response time (timestamp in millis) per request
+        // - we filter the list to only include those operations related to the WatchTower
+        // - we wait until all such responses have arrived, and are older than 5 seconds
+
```

Comment:
Why are we waiting 5 seconds? Is it to let the node publish the closing transaction in case it is needed?

What happens if electrum detects an issue but we don't have time to retaliate? Will the user be notified anyway, or do we hope that the node will have time to close the channels next time the job runs?

Reply:
Good question. I meant to ask about this.



The ElectrumClient marks the request and completed as soon as the JSON response arrives. But then the response needs to be processed by the channel. Which may result in several actions being queued. Such as storing a new state in the database, and then broadcasting a couple transactions.

I assume that the timer can fire while another coroutine is busy. Such as saving the state to the database. So I'm also assuming that some kind of delay needs to be added here.

But I'm not 100% certain about this. And I'm also not sure what a good wait time should be. Probably 5 seconds is too long.

Reply:
> What happens if electrum detects an issue but we don't have time to retaliate?

I think we can break it down into 2 scenarios:
1. `NotifyUpToDate` fires
    In this scenario, the node reacted appropriately. So either nothing needed to be done, or a commit was discovered and one or more transactions were broadcast. We schedule another background task to execute in a few days.
2. `NotifyUpToDate` does NOT fire
    Probably because the connection was slow, or electrum server was slow. Or because the OS decided to terminate the background task for other reasons. (Low battery mode?) We schedule another background task to execute in a few hours.

Either way, there's some kind of cleanup code that runs when the background task is finished. It will know whether the task finished successfully (i.e. via `NotifyUpToDate` fired). It should also be able to detect whether any channels were closed, and why they were closed. So we should be able to switch on this, and display appropriate notifications.

(The logic for this will probably be in phoenix-shared and/or the native layer.)